### PR TITLE
Add Drag and Drop file upload feature

### PR DESCRIPTION
**Pull request by okucheruk**
This PR add drag and drop file upload feature.

**Highlight**
- Upload drag zone is linked to file list.

### CODE REVIEW

File: app/page/filebrowser/FileBrowserPage.js [hunk 1]
```diff
@@ -44,10 +51,94 @@ export default class FileBrowserPage {
    * @override
    */
   _render(parentElement) {
-    new DivContainerBuilder('div')
-        .withParent(this._parentElement)
-        .withClassName('container page-container')
-        .withInnerHTML('SomeText<hr><a href="#/auth">Login</a>')
-        .build();
+    const form =
```
Comment (dkalinin) [inline app/page/filebrowser/FileBrowserPage.js]: Please rename.

File: app/page/filebrowser/FileBrowserPage.js [hunk 2]
```diff
@@ -44,10 +51,94 @@ export default class FileBrowserPage {
...
+    return this;
+  }
+
+  onUploadStart(callback) {
+    this._fileList.subscribeToEvent('drop', (event) => {
+      event.stopPropagation();
+      event.preventDefault();
+
+      this._fileList.removeClass('drop-zone');
+      const files = Object.values(event.dataTransfer.files);
```
Comment (dkalinin) [inline app/page/filebrowser/FileBrowserPage.js]: new Array(event.dataTransfer.files)

File: app/page/filebrowser/FileBrowserPage.js [hunk 3]
```diff
@@ -44,10 +51,94 @@ export default class FileBrowserPage {
...
+    return this;
+  }
+
+  _handlesFileListStateChange(newFileList) {
+
+    this._fileList.clearMarkup();
+
+    new FileBrowserListHeader(this._fileList.getNode());
+
+    newFileList.forEach((file) => {
```
Comment (dkalinin) [inline app/page/filebrowser/FileBrowserPage.js]: Let's create FileList component.

File: index.html [hunk 4]
```diff
@@ -5,7 +5,9 @@
     <meta http-equiv="X-UA-Compatible" content="IE=edge" />
     <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
     <title>Title</title>
-    <link rel="stylesheet" href="styles.css">
+<!--    <link rel="stylesheet" href="app/css/custom-styles.css">-->
+    <link rel="stylesheet" href="app/lib/bootstrap/css/bootstrap.min.css">
+    <link rel="stylesheet" href="app/css/bootstrap-styles.css">
```
Comment (dkalinin) [inline index.html]: Let's import bootstrap from node_modules.

File: app/page/filebrowser/mutation/UploadFileStatusMutation.js [hunk 5]
```diff
@@ -12,22 +12,25 @@
...
+import Mutation from '../../../lib/turbine/service/store/Mutation.js';
 
 /**
- * The UploadFileAction class represents file upload action.
+ * The UploadFileMutation class represents uploadFile state mutation.
  *
  * @author Oleksii Kucheruk
  */
-export default class UploadFileAction extends Action {
+export default class UploadFileStatusMutation extends Mutation {
```
Comment (dkalinin) [inline app/page/filebrowser/mutation/UploadFileStatusMutation.js]: Let's not use `status` for boolean fields.

File: app/page/filebrowser/action/UploadAction.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+   *
+   * @param {Store} store The storing service of application.
+   * @override
+   */
+  accept(store) {
+    store.commit(new UploadFileStatusMutation(true));
+    store.serviceFactory().newFileService().uploadFile(this._files, this._targetFolder)
+        .then((files) => {
+          // fileList is a dev simulation, use files in production.
+          const fileList = store.getState('fileList').concat(files);
```
Comment (dkalinin) [inline app/page/filebrowser/action/UploadAction.js]: fileList should contain models of files.

File: app/page/filebrowser/action/UploadAction.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+import UploadFileStatusMutation from '../mutation/UploadFileStatusMutation.js';
+import FileListMutation from '../mutation/FileListMutation.js';
+
+/**
+ * The UploadFileAction class represents file upload action.
+ *
+ * @author Oleksii Kucheruk
+ */
+export default class UploadAction extends Action {
+  constructor(files, targetFolder) {
```
Comment (dkalinin) [inline app/page/filebrowser/action/UploadAction.js]: Add JS doc.

File: app/page/filebrowser/action/UploadAction.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  /**
+   * Commits to application state store a new user file upload mutation of property.
+   *
+   * @param {Store} store The storing service of application.
+   * @override
+   */
+  accept(store) {
+    store.commit(new UploadFileStatusMutation(true));
+    store.serviceFactory().newFileService().uploadFile(this._files, this._targetFolder)
+        .then((files) => {
```
Comment (dkalinin) [inline app/page/filebrowser/action/UploadAction.js]: Please handle error response.

File: app/page/filebrowser/FileBrowserPage.js (hunk 2 above)
Comment (okucheruk) [reply app/page/filebrowser/FileBrowserPage.js]: `const files = Array.from(event.dataTransfer.files);`

File: app/page/filebrowser/FileBrowserPage.js (hunk 3 above)
Comment (okucheruk) [reply app/page/filebrowser/FileBrowserPage.js]: Refactored.

File: index.html (hunk 4 above)
Comment (okucheruk) [reply index.html]: Bootstrap 3.4.1 added as node_module.

File: app/page/filebrowser/action/UploadAction.js (hunk 6 above)
Comment (okucheruk) [reply app/page/filebrowser/action/UploadAction.js]: fileList already contain model returned by REST API.
Check the simplified data mock pushed to fileList array:
```
    uploadFile(files, parentFolderId) {
    /*
        return this._httpRequestWorker
            .post('https://doe.cloud/storemy/api/v1/filebrowser/upload/' + parentFolderId, file);
    */

    return Mocker.promisedResult(() => {
      const fileList = [];
      let id = 3;
      files.forEach((file) => {
        fileList.push(
            {
              id: id++,
              name: file.name,
              owner: 'Oleksii Kucheruk',
              modified: file.lastModifiedDate,
              size: file.size,
            });
      });
      return fileList;
    });
  }
```

File: app/page/filebrowser/FileBrowserPage.js (hunk 1 above)
Comment (okucheruk) [reply app/page/filebrowser/FileBrowserPage.js]: Renamed to `fileBrowserPage`

File: app/page/filebrowser/mutation/UploadFileStatusMutation.js (hunk 5 above)
Comment (okucheruk) [reply app/page/filebrowser/mutation/UploadFileStatusMutation.js]: Refactored to `UploadProgressMutation`

File: app/page/filebrowser/action/UploadAction.js (hunk 7 above)
Comment (okucheruk) [reply app/page/filebrowser/action/UploadAction.js]: Added.

File: index.html (hunk 4 above)
Comment (okucheruk) [reply index.html]: Imported.

File: app/page/filebrowser/mutation/UploadFileStatusMutation.js (hunk 5 above)
Comment (okucheruk) [reply app/page/filebrowser/mutation/UploadFileStatusMutation.js]: Replaced with `UploadProgressMutation`.

File: app/page/filebrowser/action/UploadAction.js (hunk 7 above)
Comment (okucheruk) [reply app/page/filebrowser/action/UploadAction.js]: Added.

File: app/page/filebrowser/action/UploadAction.js (hunk 8 above)
Comment (okucheruk) [reply app/page/filebrowser/action/UploadAction.js]: Simple error handling added for now.

File: app/page/filebrowser/FileBrowserPage.js (hunk 1 above)
Comment (okucheruk) [reply app/page/filebrowser/FileBrowserPage.js]: Renamed to `fileBrowserPage`.

File: app/page/filebrowser/FileBrowserPage.js (hunk 2 above)
Comment (okucheruk) [reply app/page/filebrowser/FileBrowserPage.js]: Correct option is `const files = Array.from(event.dataTransfer.files);`

File: app/page/filebrowser/FileBrowserPage.js (hunk 3 above)
Comment (okucheruk) [reply app/page/filebrowser/FileBrowserPage.js]: Created `FileBrowserList`.

**Comment by okucheruk** [comment]
@dkalinin, fixed, PTAL.

File: app/page/filebrowser/action/UploadAction.js [hunk 9]
```diff
@@ -35,13 +41,18 @@ export default class UploadAction extends Action {
...
-    store.commit(new UploadFileStatusMutation(true));
-    store.serviceFactory().newFileService().uploadFile(this._files, this._targetFolder)
+    store.commit(new UploadProgressMutation(true));
+
+    store.serviceFactory().fileService()
+        .upload(this._files, this._targetFolderId)
         .then((files) => {
           // fileList is a dev simulation, use files in production.
-          const fileList = store.getState('fileList').concat(files);
+          const fileList = store.getState('files').concat(files);
```
Comment (dkalinin) [inline app/page/filebrowser/action/UploadAction.js]: Let's call Fetch files action.

File: app/page/filebrowser/action/UploadAction.js [hunk 10]
```diff
@@ -35,13 +41,18 @@ export default class UploadAction extends Action {
...
+          console.log(fileList);
           store.commit(new FileListMutation(fileList));
-        });
-    store.commit(new UploadFileStatusMutation(false));
+        })
+        .catch((error) => console.log(error));
+
+    store.commit(new UploadProgressMutation(false));
```
Comment (dkalinin) [inline app/page/filebrowser/action/UploadAction.js]: Should be in async handler.

File: app/page/filebrowser/component/FileBrowserList.js [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+   * @param {Store} store The application storing service.
+   */
+  constructor(parentElement, store) {
+    this._store = store;
+    this._store.subscribeToEvent('files.stateChange', (state) => {
+      this._updateFileList(state.files);
+    });
+
+    this._fileList =
+        new DivContainerBuilder()
```
Comment (dkalinin) [inline app/page/filebrowser/component/FileBrowserList.js]: Let's have rendering and component state separated.

File: app/page/filebrowser/component/FileBrowserList.js [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+  /**
+   * Represents FileBrowserList component.
+   *
+   * @param {HTMLElement} parentElement The root element or node for the this element.
+   * @param {Store} store The application storing service.
+   */
+  constructor(parentElement, store) {
+    this._store = store;
+    this._store.subscribeToEvent('files.stateChange', (state) => {
+      this._updateFileList(state.files);
```
Comment (dkalinin) [inline app/page/filebrowser/component/FileBrowserList.js]: Let's update private property here which leads to rendering.

File: app/page/filebrowser/component/FileBrowserListItem.js [hunk 13]
```diff
@@ -17,24 +17,41 @@ import DivContainerBuilder from '../../../lib/nohtml/element/DivContainerBuilder
...
+    this._render(fileModel);
+  }
+
+  /**
+   * Renders component.
+   *
+   * @param {Object} fileModel The file model.
+   * @private
+   */
+  _render(fileModel) {
```
Comment (dkalinin) [inline app/page/filebrowser/component/FileBrowserListItem.js]: Let's not have any params in render.

**Pull request merged**